File: code/go/ch21/src/luago/compiler/lexer/token.go
package lexer

// token kind
const (
	TOKEN_EOF         = iota           // end-of-file
	TOKEN_VARARG                       // ...
	TOKEN_SEP_SEMI                     // ;
	TOKEN_SEP_COMMA                    // ,
	TOKEN_SEP_DOT                      // .
	TOKEN_SEP_COLON                    // :
	TOKEN_SEP_LABEL                    // ::
	TOKEN_SEP_LPAREN                   // (
	TOKEN_SEP_RPAREN                   // )
	TOKEN_SEP_LBRACK                   // [
	TOKEN_SEP_RBRACK                   // ]
	TOKEN_SEP_LCURLY                   // {
	TOKEN_SEP_RCURLY                   // }
	TOKEN_OP_ASSIGN                    // =
	TOKEN_OP_MINUS                     // - (sub or unm)
	TOKEN_OP_WAVE                      // ~ (bnot or bxor)
	TOKEN_OP_ADD                       // +
	TOKEN_OP_MUL                       // *
	TOKEN_OP_DIV                       // /
	TOKEN_OP_IDIV                      // //
	TOKEN_OP_POW                       // ^
	TOKEN_OP_MOD                       // %
	TOKEN_OP_BAND                      // &
	TOKEN_OP_BOR                       // |
	TOKEN_OP_SHR                       // >>
	TOKEN_OP_SHL                       // <<
	TOKEN_OP_CONCAT                    // ..
	TOKEN_OP_LT                        // <
	TOKEN_OP_LE                        // <=
	TOKEN_OP_GT                        // >
	TOKEN_OP_GE                        // >=
	TOKEN_OP_EQ                        // ==
	TOKEN_OP_NE                        // ~=
	TOKEN_OP_LEN                       // #
	TOKEN_OP_AND                       // and
	TOKEN_OP_OR                        // or
	TOKEN_OP_NOT                       // not
	TOKEN_KW_BREAK                     // break
	TOKEN_KW_DO                        // do
	TOKEN_KW_ELSE                      // else
	TOKEN_KW_ELSEIF                    // elseif
	TOKEN_KW_END                       // end
	TOKEN_KW_FALSE                     // false
	TOKEN_KW_FOR                       // for
	TOKEN_KW_FUNCTION                  // function
	TOKEN_KW_GOTO                      // goto
	TOKEN_KW_IF                        // if
	TOKEN_KW_IN                        // in
	TOKEN_KW_LOCAL                     // local
	TOKEN_KW_NIL                       // nil
	TOKEN_KW_REPEAT                    // repeat
	TOKEN_KW_RETURN                    // return
	TOKEN_KW_THEN                      // then
	TOKEN_KW_TRUE                      // true
	TOKEN_KW_UNTIL                     // until
	TOKEN_KW_WHILE                     // while
	TOKEN_IDENTIFIER                   // identifier
	TOKEN_NUMBER                       // number literal
	TOKEN_STRING                       // string literal
	TOKEN_OP_UNM      = TOKEN_OP_MINUS // unary minus
	TOKEN_OP_SUB      = TOKEN_OP_MINUS
	TOKEN_OP_BNOT     = TOKEN_OP_WAVE
	TOKEN_OP_BXOR     = TOKEN_OP_WAVE
)

var keywords = map[string]int{
	"and":      TOKEN_OP_AND,
	"break":    TOKEN_KW_BREAK,
	"do":       TOKEN_KW_DO,
	"else":     TOKEN_KW_ELSE,
	"elseif":   TOKEN_KW_ELSEIF,
	"end":      TOKEN_KW_END,
	"false":    TOKEN_KW_FALSE,
	"for":      TOKEN_KW_FOR,
	"function": TOKEN_KW_FUNCTION,
	"goto":     TOKEN_KW_GOTO,
	"if":       TOKEN_KW_IF,
	"in":       TOKEN_KW_IN,
	"local":    TOKEN_KW_LOCAL,
	"nil":      TOKEN_KW_NIL,
	"not":      TOKEN_OP_NOT,
	"or":       TOKEN_OP_OR,
	"repeat":   TOKEN_KW_REPEAT,
	"return":   TOKEN_KW_RETURN,
	"then":     TOKEN_KW_THEN,
	"true":     TOKEN_KW_TRUE,
	"until":    TOKEN_KW_UNTIL,
	"while":    TOKEN_KW_WHILE,
}

File: code/java/ch18/src/main/java/com/github/zxh0/luago/compiler/lexer/Token.java
package com.github.zxh0.luago.compiler.lexer;

import lombok.Getter;

import java.util.HashMap;
import java.util.Map;

@Getter
public class Token {

    static final Map<String, TokenKind> keywords = new HashMap<>();
    static {
        keywords.put("and",      TokenKind.TOKEN_OP_AND);
        keywords.put("break",    TokenKind.TOKEN_KW_BREAK);
        keywords.put("do",       TokenKind.TOKEN_KW_DO);
        keywords.put("else",     TokenKind.TOKEN_KW_ELSE);
        keywords.put("elseif",   TokenKind.TOKEN_KW_ELSEIF);
        keywords.put("end",      TokenKind.TOKEN_KW_END);
        keywords.put("false",    TokenKind.TOKEN_KW_FALSE);
        keywords.put("for",      TokenKind.TOKEN_KW_FOR);
        keywords.put("function", TokenKind.TOKEN_KW_FUNCTION);
        keywords.put("goto",     TokenKind.TOKEN_KW_GOTO);
        keywords.put("if",       TokenKind.TOKEN_KW_IF);
        keywords.put("in",       TokenKind.TOKEN_KW_IN);
        keywords.put("local",    TokenKind.TOKEN_KW_LOCAL);
        keywords.put("nil",      TokenKind.TOKEN_KW_NIL);
        keywords.put("not",      TokenKind.TOKEN_OP_NOT);
        keywords.put("or",       TokenKind.TOKEN_OP_OR);
        keywords.put("repeat",   TokenKind.TOKEN_KW_REPEAT);
        keywords.put("return",   TokenKind.TOKEN_KW_RETURN);
        keywords.put("then",     TokenKind.TOKEN_KW_THEN);
        keywords.put("true",     TokenKind.TOKEN_KW_TRUE);
        keywords.put("until",    TokenKind.TOKEN_KW_UNTIL);
        keywords.put("while",    TokenKind.TOKEN_KW_WHILE);
    }

    private final int line;
    private final TokenKind kind;
    private final String value;

    public Token(int line, TokenKind kind, String value) {
        this.line = line;
        this.kind = kind;
        this.value = value;
    }

}
